src/number.ml

```diff
@@ -208,3 +208,206 @@ module Make (Impl : Snark_intf.Basic) = struct
     let%bind less = x < y in
     if_ less ~then_:y ~else_:x
 end
+
+module Run = struct
+  module Make (Impl : Snark_intf.Run_basic) = struct
+    open Impl
+
+    type t =
+      { upper_bound: Bignum_bigint.t
+      ; lower_bound: Bignum_bigint.t
+      ; var: Field.t
+      ; bits: Boolean.var list option }
+
+    let two_to_the n =
+      let rec go acc i =
+        if i <= 0 then acc else go (Field.Constant.add acc acc) (i - 1)
+      in
+      go Field.Constant.one n
+
+    let to_bits {var; bits; upper_bound; lower_bound= _} =
```

Comment:
Can we use the monadic versions to implement all these? It's a bit sad to duplicate all this code.

Reply:
Done